### ISSUE TITLE
How to share code between platforms?

### ISSUE THREAD

**Issue by FokkeZB** [hansemannn/hyperloop-module-template#3]
After all, a major benefit of HL is using JS on all platforms.

Maybe we should approach HL modules more as CommonJS modules, less like Titanium classic modules and use the existing platform-specific folders that Titanium supports.

What if HL modules would have a structure like:

```
./src/index.js      # main cross-platform entry for require('ti-test')
./src/ios/index.js  # platform-specific entry
```

Which during build would be injected as if you would have:

```
Resources/ti-test/index.js
Resources/ti-test/ios/index.js
```

Which would result in getting:

```
build/iphone/assets/ti-test/index_js # source of app/lib/ti-test/ios/index.js
```

For assets it could work like Alloy widgets that use the [WPATH](https://github.com/appcelerator/alloy/blob/master/Alloy/template/wpath.js) trick:

```
./assets/images/foo/image.png
./assets/ios/images/foo/image.png
```

Would be injected as if they were at:

```
Resources/images/ti-test/foo/image.png
Resources/ios/images/ti-test/foo/image.png
```

And in the module you would have a similar helper to translate the path via `MPATH('images/foo/image.png')`, which would use `module.id` to resolve.

**Comment by hansemannn** [comment]
Good point! I guess it's a design decision. As all hyperloop modules are commonJS-modules, we should be able to. Can you submit a PR on how you would design the interfaces, thinking of commonJS best practices? Don't want to let it like it currently is (which is too much like native modules as you said).
